File: haskell/kindergarten-garden/Garden.hs
module Garden where

import Data.List
import Data.List.Split(chunksOf)

data Plant = Radishes | Violets | Clover | Grass deriving (Eq, Show)
type Child = String
type Garden = [(Child, [Plant])]

letterToPlant :: Char -> Plant
letterToPlant letter = case letter of
    'R' -> Radishes
    'V' -> Violets
    'C' -> Clover
    'G' -> Grass
    otherwise -> error "Porra, vey"

garden :: [Child] -> String -> Garden
garden children cells = zip (sort children) (splitCells cells)
    where
        splitCells = map concat . transpose . map
            (chunksOf 2 . map letterToPlant) . lines

defaultGarden :: String -> Garden
defaultGarden = garden ["Alice", "Bob", "Charlie", "David", "Eve", "Fred",
                "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"]

lookupPlants :: Child -> Garden -> [Plant]
lookupPlants child = snd . head . filter ((==child) . fst)

File: python/kindergarten-garden/garden.py
class Garden(object):
    default_students = "Alice Bob Charlie David Eve Fred \
        Ginny Harriet Ileana Joseph Kincaid Larry".split()

    def __init__(self, garden, students=default_students):
        rows = garden.splitlines()
        if students != Garden.default_students:
            students.sort()

        self.cups = {}

        for row in rows:
            plants = iter(row)
            for student in students:
                try:
                    if student not in self.cups:
                        self.cups[student] = []
                    self.cups[student].append(next(plants))
                    self.cups[student].append(next(plants))
                except:
                    pass
        for student in students:
            cups = map(Garden.charToPlant, self.cups[student])
            self.cups[student] = list(cups)

    def plants(self, kid):
        return self.cups[kid]

    def charToPlant(char):
        if char == 'V':
            return 'Violets'
        if char == 'R':
            return 'Radishes'
        if char == 'G':
            return 'Grass'
        if char == 'C':
            return 'Clover'
